### ISSUE TITLE
Implement New Training Loop

### ISSUE THREAD

**Issue by zparis99** [hassonlab/podcast-benchmark#15]
The training loop for this approach needs to be more complex than our existing one. I think at this point it makes sense to split up the actual training code into two separate functions that we share a lot of util code between. This task involves:

1. Breaking up the existing training loop into useful util functions (preferably with tests since this provides a good opportunity to do so)
2. Rewriting our existing training loop using those utils
3. Writing a new training loop which implements the warmup and training phase from the paper while also maximally reusing the util code.
4. Adding an if-statement before calling the training loop to split which loop is run depending on the task. All simple tasks can use the old training code. This new one will just have its own.

While we could try to generalize the current training loop to support this new code without any if-statements it would quickly turn into a mess as we'd have to add a few new function registries which we'd only use for a single task anyway. At this point we may as well keep it simple and separate with as much code reuse as we can manage to strike a good middle ground.
